crates/sui-core/src/consensus_adapter.rs

```diff
@@ -226,48 +268,101 @@ impl ConsensusAdapter {
-        // In the future we can actually use information about consensus rounds instead of this delay
-        const DELAY_STEP: Duration = Duration::from_secs(7);
-        (
-            DELAY_STEP * std::cmp::min(position, MAX_DELAY_MUL) as u32,
-            position,
-        )
+        low_scoring_authorities: Vec<&AuthorityName>,
+    ) -> usize {
+        let positions = order_validators_for_submission(committee, tx_digest);
+
+        self.check_submission_wrt_connectivity(ourselves, positions, low_scoring_authorities)
+    }
+
+    /// This function runs the following algorithm to decide whether or not to submit a transaction
+    /// to consensus.
+    ///
+    /// It takes in a deterministic list that represents positions of all the authorities.
+    /// The authority in the first position will be responsible for submitting to consensus, and
+    /// so we check if we are this validator, and if so, return true.
+    ///
+    /// If we are not in that position, we check our connectivity to the authority in that position.
+    /// If we are connected to them, we can assume that they are operational and will submit the transaction.
+    /// If we are not connected to them, we assume that they are not operational and we will not rely
+    /// on that authority to submit the transaction. So we shift them out of the first position, and
+    /// run this algorithm again on the new set of positions.
+    ///
+    /// This can possibly result in a transaction being submitted twice if an authority sees a false
+    /// negative in connectivity to another, such as in the case of a network partition.
+    ///
+    /// Recursively, if the authority further ahead of us in the positions is a low performing authority, we will
+    /// move our positions up one, and submit at the same time. This allows low performing
+    /// node a chance to participate in consensus and redeem their scores while maintaining performance.
+    fn check_submission_wrt_connectivity(
+        &self,
+        ourselves: &AuthorityName,
+        positions: Vec<AuthorityName>,
+        low_scoring_authorities: Vec<&AuthorityName>,
+    ) -> usize {
+        let filtered_positions: Vec<AuthorityName> = positions
+            .into_iter()
+            .filter(|authority| {
+                self.connection_monitor_status
+                    .check_connection(authority)
+                    .unwrap_or(ConnectionStatus::Disconnected)
+                    == ConnectionStatus::Connected
+            })
+            .filter(|authority| {
+                // if we are a low scoring authority, we do not filter out ourselves from the list,
+                // nor do we move forward in line
+                // if we are a high scoring authority, we will co-submit with any low scoring authorities
+                // in front of us
+                !low_scoring_authorities.contains(&authority)
+                    || low_scoring_authorities.contains(&ourselves)
```

Comment:
I considered this actually and that is slightly different. What that would acheive is for all the low scoring nodes to remove all the other low scoring nodes but themselves, which would result in more of the low scoring nodes to submit the same transaction sometimes.

Reply:
> What that would acheive is for all the low scoring nodes to remove all the other low scoring nodes but themselves

Ok but if we have the code like this:
```
!low_scoring_authorities.contains(&authority)
                    || low_scoring_authorities.contains(&ourselves)
```

then if _we_ are a low scoring authority, we'll always include everyone else in the final set - irrespective of whether we consider them low scoring authority or not - so practically it would be like disabling the scoring mechanism, right? Is this intended?

Reply:
Yes, that is intentional, in the case a tx gets mapped like this:
[low_scoring_a1, low_scoring_a2, a3]
It would result in low_scoring_a1 and a3 to submit the tx since low_scoring_a2 did not filter out low_scoring_a1. I was aiming to avoid an amplification where all 3 submit the same tx.